naming/src/main/java/com/alibaba/nacos/naming/core/ServiceManager.java

```diff
@@ -466,24 +466,27 @@ public void createServiceIfAbsent(String namespaceId, String serviceName, boolea
             throws NacosException {
         Service service = getService(namespaceId, serviceName);
         if (service == null) {
-            
-            Loggers.SRV_LOG.info("creating empty service {}:{}", namespaceId, serviceName);
-            service = new Service();
-            service.setName(serviceName);
-            service.setNamespaceId(namespaceId);
-            service.setGroupName(NamingUtils.getGroupName(serviceName));
-            // now validate the service. if failed, exception will be thrown
-            service.setLastModifiedMillis(System.currentTimeMillis());
-            service.recalculateChecksum();
-            if (cluster != null) {
-                cluster.setService(service);
-                service.getClusterMap().put(cluster.getName(), cluster);
-            }
-            service.validate();
-            
-            putServiceAndInit(service);
-            if (!local) {
-                addOrReplaceService(service);
+            synchronized (this) {
```

Comment:
lock in this may affect the performance.
I think fix like following will be better
```
    public void putService(Service service) {
        if (!serviceMap.containsKey(service.getNamespaceId())) {
            synchronized (putServiceLock) {
                if (!serviceMap.containsKey(service.getNamespaceId())) {
                    serviceMap.put(service.getNamespaceId(), new ConcurrentHashMap<>(16));
                }
            }
        }
        *serviceMap.get(service.getNamespaceId()).putIfAbsent(service.getName(), service);
    }

    private void putServiceAndInit(Service service) throws NacosException {
        putService(service);
        *service = getService(service.getNamespaceId(), service.getName());
        service.init();
        consistencyService.listen(KeyBuilder.buildInstanceListKey(service.getNamespaceId(), service.getName(), true), service);
        consistencyService.listen(KeyBuilder.buildInstanceListKey(service.getNamespaceId(), service.getName(), false), service);
        Loggers.SRV_LOG.info("[NEW-SERVICE] {}", service.toJSON());
    }
```

What's your think about

Reply:
you are right,  "synchronized (this)" in this method will affect the performance, your plan is better